File: src/lambdas.py
import expr

def isseq (o):
    return isinstance (o, tuple) or isinstance (o, list)

def isvar (o):
    return isinstance (o, str) and o.startswith('_') and not o.startswith('__')

def ispat (o):
    return isinstance (o, str) and o.startswith('__')

def islambda (o):
    return isseq (o) and len (o) == 3 and isvar (o[0]) and o[1] == ':'

def walk (o, prefix=()):
    if islambda (o):
        yield (o, prefix)
        yield from walk (o[2], prefix + (2,))
    elif isseq (o):
        yield (o, prefix)
        for i, sub in enumerate (o):
            yield from walk (sub, prefix + (i,))
    else:
        yield (o, prefix)

def _all_path_indices (term, prefix):
    yield tuple (prefix)
    if isseq (term):
        for (i, t) in enumerate (term):
            prefix.append (i)
            yield from _all_path_indices (t, prefix)
            prefix.pop ()

def all_path_indices (term):
    return list (_all_path_indices (term, []))

def sub_by_path (term, path):
    if not path:
        return term
    else:
        return sub_by_path (term[path[0]], path[1:])

def replace_by_path (term, path, target):
    if not path:
        return target
    else:
        ans = list (term)
        ans[path[0]] = replace_by_path (ans[path[0]], path[1:], target)
        return tuple (ans)

def hashead (obj, head):
    return isseq (obj) and obj and obj[0] == head

def all_symbols (o):
    if isseq (o):
        return {sym for subo in o for sym in all_symbols (subo)}
    else:
        return {o}

def all_vars (o):
    if isseq (o):
        return {x for x in all_symbols (o) if isvar (x)}
    else:
        return {o}

def free_vars (o):
    if islambda (o):
        return free_vars (o[2]) - {o[0]}
    elif isseq (o):
        return {var for subo in o for var in free_vars (subo)}
    else:
        return {o} if isvar (o) else set ()

def free_vars_and_pats (o):
    if islambda (o):
        return free_vars_and_pats (o[2]) - {o[0]}
    elif isseq (o):
        return {var for subo in o for var in free_vars_and_pats (subo)}
    else:
        return {o} if isvar (o) or ispat (o) else set ()

def genvar (original, avoid):
    while original in avoid:
        original = original + "+"
    return original

def lambda_b_reduce (lam, arg, avoid_binding = ()):
    if not islambda (lam):
        raise ValueError (f'Bad lambda {lam}')
    var = lam[0]
    return lambda_replace (lam[2], var, arg, avoid_binding)

# A note on variable capture.
# For EFA, the usual replacement process is safe
#     and capture detection may generally be unnecessary,
#     save for the cases where `arg` contains a bound variable
#     that conflicts with a free variable.
# Another possible case exists when a pattern of the form
#     (__f _x) is realized, which invokes replacement.
def lambda_replace (exp, var, arg, avoid_binding = ()):
    if islambda (exp):
        # Lambda. Beware of variable capture.
        innervar = exp[0]
        innerexp = exp[2]
        if innervar in all_vars (arg):
            newvar = genvar (original = innervar,
                               avoid = all_symbols (arg) .union (all_symbols (exp)) .union (set (avoid_binding)))
            innerexp = lambda_replace (innerexp, innervar, newvar, ())
            innervar = newvar
        new_innerexp = lambda_replace (innerexp, var, arg, avoid_binding)
        return (innervar, ':', new_innerexp)
    elif isseq (exp):
        return tuple (
            lambda_replace (subexp, var, arg, avoid_binding)
            for subexp in exp
        )
    else:
        # Atomic.
        if exp == var: return arg
        else:          return exp

def rich_lambda_replace (exp, f, arg, avoid_binding = ()):
    if islambda (exp):
        innervar = exp[0]
        innerexp = exp[2]
        if innervar in all_vars (arg):
            newvar = genvar (original = innervar,
                               avoid = all_symbols (arg) .union (all_symbols (exp)) .union (set (avoid_binding)))
            innerexp = lambda_replace (innerexp, innervar, newvar, ())
            innervar = newvar
        new_innerexp = rich_lambda_replace (innerexp, f, arg, avoid_binding)
        return (innervar, ':', new_innerexp)
    elif isseq (exp):
        exp = tuple (rich_lambda_replace (subexp, f, arg, avoid_binding)
                     for subexp in exp)
        # Detect (f x)
        if exp[0] == f and len (exp) == 2:
            exp = lambda_b_reduce (arg, exp[1], avoid_binding)
        return exp
    else:
        return exp

def lambda_eq (a, b):
    # Judges if `a` and `b` are the same lambda expression.
    return lambda_normal (a) == lambda_normal (b)

def lambda_normal (lam, reserved_prefix='_!R'):
    # Returns the "normal form" of a lambda expression.
    lam, _ = _lambda_normal (lam, countfrom = 0, varmap = {}, reserved_prefix = reserved_prefix)
    return lam

def _lambda_normal (lam, countfrom, varmap, reserved_prefix='_!R'):
    if islambda (lam):
        (var, _, sub) = lam
        rname = reserved_prefix + str (countfrom)
        subvarmap = {**varmap}
        subvarmap[var] = rname
        sub, tail = _lambda_normal (sub, countfrom + 1, subvarmap)
        return (rname, ':', sub), tail
    elif isseq (lam):
        ans = []
        for sub in lam:
            resub, countfrom = _lambda_normal (sub, countfrom, varmap)
            ans.append (resub)
        return (tuple (ans), countfrom)
    else:
        # Atomic.
        return varmap.get (lam, lam), countfrom

def lambda_valid (lam, avoid = ()):
    # Returns if `lam` is a valid lambda expression.
    if islambda (lam):
        (var, _, sub) = lam
        return var not in avoid and lambda_valid (sub, avoid + (var,))
    elif isseq (lam):
        return all (lambda_valid (sub, avoid)
                    for sub in lam)
    else:
        # Atomic.
        return True

def bindings_by (lam, path, prefix = ()):
    if not path:
        return prefix
    elif islambda (lam):
        return bindings_by (lam[path[0]], path[1:], prefix + (lam[0],))
    else:
        return bindings_by (lam[path[0]], path[1:], prefix)

def is_redux (lam):
    return len (lam) == 2 and islambda (lam[0])

# Finds all abstraction opportunities in `term`
# Try entering:
#     (_x : (_y : (= (+ _x _y) (+ _y _x) (_z : (+ _x _y)))))
#     (_x : (_y : (= (+ _x _y) (+ _y _x) (_x : (+ _x _y)))))
def abstractions (term):
    nf = lambda_normal (term)
    maps = {}
    for nfo, path in walk (nf):
        o = sub_by_path (term, path)
        subnf = lambda_normal (nfo, reserved_prefix='_!S')
        maps.setdefault (subnf, [])
        maps[subnf].append (path)
    for elem, paths in maps.items ():
        print ('Term', elem, 'appeared in:')
        for p in paths:
            print ('   ', p, 'as', sub_by_path (term, p))
        print ()

if __name__ == '__main__':
    try:
        while True:
            line = input ()
            row = expr.parse (line)
            abstractions (row)
    except EOFError:
        exit (0)

File: js/lambdas.js
// Anything match- and lambda-related.

// TODO:
// * lambda expressions
// * alpha-conversion
// * beta-reduction
// * beta(!)-expansion

// Borrow from lambdas.py

function isLambda(sexp) {
  return simpleMatch([':', '_A', '_B'], sexp).success && isVar(sexp[1]);
}

function isRedux(sexp) {
  return simpleMatch([[':', '_A', '_B'], '_C'], sexp).success;
}

function walkSexp(sexp, prefixQ) {
  const prefix = prefixQ || [];
  if (isLambda(sexp)) {
    return [[sexp, prefix], ...walkSexp(sexp[2], prefix.concat([2]))];
  } else if (isList(sexp)) {
    const ans = [[sexp, prefix]];
    for (let i = 0; i < sexp.length; i++) {
      ans.push(... walkSexp(sexp[i], prefix.concat([i])));
    }
    return ans;
  } else {
    return [[sexp, prefix]];
  }
}

function subByPath(sexp, path) {
  if (eq([], path)) return sexp;
  return subByPath(sexp[path[0]], path.slice(1));
}

function replaceByPath(sexp, path, target) {
  if (eq([], path)) return target;

  const ans = [...sexp];
  ans[path[0]] = replaceByPath(ans[path[0]], path.slice(1), target);
  return ans;
}

function getAllAtoms(sexp) {
  if (isAtomic(sexp)) return [sexp];

  const ans = [];
  for (const sub of sexp) ans.push(...getAllAtoms(sub));
  return ans;
}

function getAllVars(sexp) {
  return getAllAtoms(sexp).filter(isVar);
}

function getFreeVars(sexp) {
  if (isLambda(sexp)) {
    const [_, varName, body] = sexp;
    const ans = new Set(getFreeVars(body));  // corrected!
    ans.delete(varName);
    return [...ans];
  } else if (isList(sexp)) {
    const ans = [];
    for (const sub of sexp) ans.push(...getFreeVars(sub));
    return ans;
  } else if (isVar(sexp)) {
    return [sexp];
  } else {
    return [];
  }
}

function genVar(avoid, ts) {
  for (let i = 0; i <= avoid.length; i++) {
    const contender = `_v${i}:${ts}`;
    if (! avoid.includes(contender)) return contender;
  }
}

function gensyms (avoid, count = 1, prefix = '#', suffix = '') {
  const avoidSet = new Set(getAllAtoms(avoid));
  const ans = [];
  for (let i = 0; i <= avoidSet.size + count; i++) {
    const contender = `${prefix}${i}${suffix}`;
    if (! avoidSet.has(contender)) {
      ans.push(contender);
      if (ans.length >= count) return ans;
    }
  }
}

if (false) {
  console.log(getAllVars(parseOne(`
      (forall (: _x:O [= _x:O (+ O _x:O)]))
    `)));
  console.log(getFreeVars(parseOne(`
      (forall (: _x:O [= _x:O (+ O _x:O)]))
    `)));
  console.log(getFreeVars(parseOne(`
      (forall (: _x:O [= _x:O (+ _y:O _x:O)]))
    `)));
  console.log(genVar(['_v0:O', '_v1:O', '_v5:O'], 'O'));
}

function lambdaReduce (lam, arg) {
  if (! isLambda(lam)) {
    return null;
  }
  const [_, v, body] = lam;
  return lambdaReplace (body, v, arg);
}

function lambdaReplace (sexp, v, arg) {
  if (isLambda (sexp)) {
    // Lambda. Beware variable capture.
    let innerVar = sexp[1], innerExp = sexp[2];
    if (eq(innerVar, v)) return sexp;  // Variable shadowing.
    if (getFreeVars(arg).includes(innerVar)) {
      // Replace.
      const newVar = genVar(
        getFreeVars(arg).concat(getAllVars(sexp)),
        typeToString(getType(innerVar)),
      );
      innerExp = lambdaReplace(innerExp, innerVar, newVar);
      innerVar = newVar;  // Order corrected!
    }
    return [':', innerVar, lambdaReplace(innerExp, v, arg)];
  } else if (isList(sexp)) {
    return sexp.map((m) => lambdaReplace(m, v, arg));
  } else {
    return isVar(sexp) && eq(sexp, v) ? arg : sexp;
  }
}

function lambdaNormal(sexp, map = null, resPrefix = '_!R', countfrom = 0n) {
  const varmap = map || new Map();
  if (isLambda(sexp)) {
    const [_, v, body] = sexp;
    const rname = resPrefix + str(countfrom);
    const subvarmap = new Map([... varmap]);
    subvarmap.set(v, rname);
    const [sub, tail] = lambdaNormal(body, subvarmap, resPrefix, countfrom + 1n);
    return [[':', rname, sub], tail];
  } else if (isList(sexp)) {
    const ans = [];
    let cf = countfrom, resub;
    for (const sub of sexp) {
      [resub, cf] = lambdaNormal(sub, varmap, resPrefix, cf);
      ans.push(resub);
    }
    return [ans, cf];
  } else {
    return [varmap.get(sexp) || sexp, countfrom];
  }
}

function lambdaEq(l1, l2) {
  const [a, _a] = lambdaNormal(l1);
  const [b, _b] = lambdaNormal(l2);
  return eq(a, b);
}

// Inverse of `lambdaReduce`. Returns a collection of possible replacements.
function lambdaExtract (term) {
  const nf = lambdaNormal (term) [0];
  const maps = new Map();

  walkSexp(nf).map((x) => console.log(str(x)))
  for (const [nfo, path] of walkSexp(nf)) {
    const sub = subByPath(term, path);
    const gist = str(lambdaNormal(nfo, new Map(), '_!S')[0]);
    if (! maps.has(gist)) {
      maps.set(gist, []);
    }
    maps.get(gist).push(path);
  }

  return maps;
}

// Returns # of reductions made.
function lambdaOneStepReduce(term) {
  if (isAtomic(term)) return [term, 0];
  if (isLambda(term)) {
    const [_, v, body] = term;
    const [newBody, reductionsMade] = lambdaOneStepReduce(body);
    return [[':', v, newBody], reductionsMade];
  }

  // Application.
  const self = [...term];
  for (var i = 0; i < self.length; i++) {
    const [newTerm, reductionsMade] = lambdaOneStepReduce(self[i]);
    self[i] = newTerm;
    if (reductionsMade > 0) {
      return [self, reductionsMade];
    } else {
      // Normal case; keep reducing.
    }
  }
  // No reductions yet; check of entire term is reduction.
  if (/*isRedux*/ self.length === 2 && isLambda(self[0])) {
    const reduced = lambdaReduce(
      self[0], self[1]
    );
    return [reduced, 1];
  }

  // No reductions
  return [self, 0];
}

function lambdaFullReduce(term, stepLimit = 100) {
  let t = term, reductionsMade = 0;
  for (var i = 0; i < stepLimit; i++) {
    [t, reductionsMade] = lambdaOneStepReduce(t);
    if (reductionsMade <= 0) break;
  }
  return t;
}

if (0) {
  const start = parseOne(`[forall [[: _p:<OP> [: _n:O [-> [_p:<OP> _n:O] [_p:<OP> [S _n:O]]]]] [: _x:O [= [+ O _x:O] O]]]]`)
  console.log(lambdaFullReduce(start));
  /*
    [forall [[: _p:<OP> [: _n:O [-> [_p:<OP> _n:O] [_p:<OP> [S _n:O]]]]] [: _x:O [= [+ O _x:O] O]]]]
    [forall [: _n:O [-> [[: _x:O [= [+ O _x:O] O]] _n:O] [[: _x:O [= [+ O _x:O] O]] [S _n:O]]]]]
    [forall [: _n:O [-> [= [+ O _n:O] O] [[: _x:O [= [+ O _x:O] O]] [S _n:O]]]]]
    [forall [: _n:O [-> [= [+ O _n:O] O] [= [+ O [S _n:O]] O]]]]
    [forall [: _n:O [-> [= [+ O _n:O] O] [= [+ O [S _n:O]] O]]]]
  */

  console.log(lambdaFullReduce(parseOne(`
      [(: _x (_x _x)) (: _x (_x _x))]
    `)))
}

if (0) {
  console.log(str(lambdaReplace(
    parseOne(`(: _x:O (: _y:O [+ _x:O _z:O]))`),
    '_z:O',
    '3',
  )))

  console.log(str(lambdaNormal(
    parseOne(`(: _x (: _x [+ _x _y _z]))`), new Map(),
  )))

  console.log(lambdaEq(
    parseOne(`(: _x (: _x [+ _x _y _z]))`),
    parseOne(`(: _x (: _u [+ _u _y _z]))`),
  ))

  console.log(lambdaEq(
    parseOne(`(: _x (: _x [+ _x _y _z]))`),
    parseOne(`(: _x (: _u [+ _u _y _x]))`),
  ))

  console.log(walkSexp(parseOne(`(: _x (: _x [+ _x _y _z]))`), []))

  console.log(lambdaExtract(parseOne(
    `([: _x (* 2 (= _x _x))] [: _x _x])`
  )))
}

// TODO: abstractions
